typescript/src/redemption.ts

```diff
@@ -0,0 +1,299 @@
+ * @param redeemerAddresses - The list of redeemer addresses
+ * @param witness - The parameter used to decide the type of the change output.
+ *                  P2WPKH if `true`, P2PKH if `false`
+ * @returns Empty promise.
+ */
+export async function redeemDeposits(
+  bitcoinClient: BitcoinClient,
+  bridge: Bridge,
+  walletPrivateKey: string,
+  mainUtxo: UnspentTransactionOutput,
+  redeemerAddresses: string[],
+  witness: boolean
+): Promise<void> {
+  const rawTransaction = await bitcoinClient.getRawTransaction(
+    mainUtxo.transactionHash
+  )
+
+  const mainUtxoWithRaw: UnspentTransactionOutput & RawTransaction = {
+    ...mainUtxo,
+    transactionHex: rawTransaction.transactionHex,
+  }
+
+  const redemptionRequests = await prepareRedemptionRequests(
+    bridge,
+    walletPrivateKey,
+    redeemerAddresses
+  )
+
+  const transaction = await createRedemptionTransaction(
+    walletPrivateKey,
+    mainUtxoWithRaw,
+    redemptionRequests,
+    witness
+  )
+
+  // Note that `broadcast` may fail silently (i.e. no error will be returned,
+  // even if the transaction is rejected by other nodes and does not enter the
+  // mempool, for example due to an UTXO being already spent).
+  await bitcoinClient.broadcast(transaction)
+}
+
+/**
+ * Creates a Bitcoin redemption transaction.
+ * The transaction will have a single input (main UTXO) and an output for each
+ * redemption request provided and a change output if the redemption requests
+ * do not consume all the Bitcoins from the main UTXO.
+ * @dev The caller is responsible for ensuring the redemption request list is
+ *      correctly formed:
+ *        - there is at least one redemption
+ *        - the `requestedAmount` in each redemption request is greater than
+ *          the sum of its `feeShare` and `treasuryFee`
+ *        - the redeemer address in each redemption request is of a standard
+ *          type (P2PKH, P2WPKH, P2SH, P2WSH).
+ * @param walletPrivateKey  - The private key of the wallet in the WIF format
+ * @param mainUtxo - The main UTXO of the wallet. Must be P2(W)PKH
+ * @param redemptionRequests - The list of redemption requests
+ * @param witness - The parameter used to decide the type of the change output.
+ *                  P2WPKH if `true`, P2PKH if `false`
+ * @returns Bitcoin redemption transaction in the raw format.
+ */
+export async function createRedemptionTransaction(
+  walletPrivateKey: string,
+  mainUtxo: UnspentTransactionOutput & RawTransaction,
+  redemptionRequests: RedemptionRequest[],
+  witness: boolean
+): Promise<RawTransaction> {
+  if (redemptionRequests.length < 1) {
+    throw new Error("There must be at least one request to redeem")
+  }
+
+  const decodedWalletPrivateKey = wif.decode(walletPrivateKey)
+
+  const walletKeyRing = new bcoin.KeyRing({
+    witness: witness,
+    privateKey: decodedWalletPrivateKey.privateKey,
+    compressed: decodedWalletPrivateKey.compressed,
+  })
```

Comment:
Let's use `createKeyRing` from `deposit.ts`. I think we can also move that function to `bitcoin.ts`. Let's also take an opportunity and use it in `deposit-sweep.ts`.